[IOTSFW-2509] Fix for broken yotta and jenkins builds

hasnainvirk opened the pull request:
This PR will fix the broken yotta and jenkins builds with uvisor

@bogdanm  @AlessandroA

bogdanm commented on ld/K64FN1M0xxx12.ld:
```diff
@@ -7,7 +7,7 @@ MEMORY
   FLASH_PROTECTION (rx) : ORIGIN = 0x00000400, LENGTH = 0x00000010
   FLASH (rx)            : ORIGIN = 0x00000410, LENGTH = 0x00100000 - 0x00000410
-  RAM (rwx)             : ORIGIN = 0x1FFF0000, LENGTH = 0x00010000 
+  RAM (rwx)             : ORIGIN = 0x1FFF0000, LENGTH = 0x00020000 
```
That doesn't look right. You're overflowing RAM into RAM2 now, what's the point of this?

teetak01 commented:
+1, at least fixed issues on mbed Client side

bogdanm commented:
@AlessandroA, can you please take a look at this?

kjbracey committed:
[IOTSFW-2509] Rework memory layout

Previous change to memory layout to avoid the 64K RAM discontinuity broke some
builds, as it restricted static data to the first 64K. Only stack and dlmalloc
heap were in the top 192K. Nanostack builds were particularly affected as the
Nanostack heap is a large bss allocation, and didn't fit.

Rework the memory allocation to better use the memory - stack and uVisor data
now go in the first 64K, with the stack auto-sized. Data, bss and dlmalloc heap
go in the 192K region.

Also fix the broken uVisor RAM layout symbols and restore the 512-byte initial
reservation (for uVisor?).

Stack limit checking also restored by putting the stack back immediately above
the secure memory.

Previous layout
---------------
1fff0000 RAM  uvisor.bss (512-byte reservation missing)
1fff2000      data, bss
1fff....      end of bss
(some unused RAM)
20000000 RAM2 stack (fixed 32K)
20008000      start of dlmalloc heap (fixed 160K)
2002ffff      end of dlmalloc heap

New layout
----------
1fff0000      Start of physical ram, 512 bytes reserved
1fff0200 RAM  uvisor.bss
1fff....      start of stack (autosized 64K - uvisor data)
1fffffff      end of stack
20000000 RAM2 data, bss
2000....      start of dlmalloc heap (autosized 192K - static data)
2002ffff      end of dlmalloc heap

hasnainvirk commented:
@kjbracey-arm , @AlessandroA , @bogdanm

kjbracey commented:
We've created a second attempt at this. Have another look.

bogdanm commented:
Looks much better now, but I'd really like @AlessandroA to look it at (or maybe @meriac if Alessandro is not available).

kjbracey commented:
I'd like to double-check what the 512 byte reservation is (was?) for, and document it.

Note that there is now a detailed commit message explaining what's happening - GitHub doesn't seem to make this obvious.

teetak01 commented:
+1 this also works for Client

ghost commented:
The description says Ram contains uvisor.bss followed by stack, but the script itself only appears to place the stack in RAM, everything else goes to RAM2. Is the description wrong or have I mis-read the script?

kjbracey commented:
Lines 59-80, untouched by this commit, are putting uvisor.bss in RAM.

meriac commented:
++, LGTM @mjs-arm @hasnainvirk

bogdanm commented:
Thanks a lot @meriac!
